src/modules/navigator/mission.cpp

```diff
@@ -1925,3 +1928,127 @@ void Mission::publish_navigator_mission_item()
 
 	_navigator_mission_item_pub.publish(navigator_mission_item);
 }
+
+
+void Mission::calculate_mission_checksums()
+{
+	mission_s mission_state = {};
+	mission_stats_entry_s stats;
+	mission_checksum_s csum{};
+
+	csum.timestamp = hrt_absolute_time();
+	csum.mission_checksum = 0;
+	csum.fence_checksum = 0;
+	csum.rally_checksum = 0;
+	csum.all_checksum = 0;
+
+	dm_lock(DM_KEY_MISSION_STATE);
+
+	if (dm_read(DM_KEY_MISSION_STATE, 0, &mission_state, sizeof(mission_s)) != sizeof(mission_s)) {
+		PX4_ERR("dataman read failure");
+		return;
+	}
+
+	dm_unlock(DM_KEY_MISSION_STATE);
+
+	dm_item_t dm_current = (dm_item_t)(_mission.dataman_id);
+
+	for (int i = 0; i < mission_state.count; i++) {
+		struct mission_item_s mission_item = {};
+		const ssize_t len = sizeof(mission_item);
+
+		if (dm_read(dm_current, i, &mission_item, len) != len) {
+			PX4_ERR("dataman read failure");
+			return;
+		}
+
+		uint8_t frame = mission_item.frame;
+		csum.mission_checksum = crc32part(&frame, sizeof(frame), csum.mission_checksum);
+		csum.mission_checksum = crc32part(reinterpret_cast<uint8_t *>(&mission_item.nav_cmd), sizeof(mission_item.nav_cmd),
+						  csum.mission_checksum);
+		uint8_t autocontinue = mission_item.autocontinue;
+		csum.mission_checksum = crc32part(&autocontinue, sizeof(autocontinue), csum.mission_checksum);
+
+		for (int j = 0; j < 4; j++) {
+			csum.mission_checksum = crc32part(reinterpret_cast<uint8_t *>(&mission_item.params[j]), sizeof(mission_item.params[j]),
+							  csum.mission_checksum);
+		}
+
+		float lat = static_cast<float>(mission_item.lat);
+		csum.mission_checksum = crc32part(reinterpret_cast<uint8_t *>(&lat), sizeof(lat), csum.mission_checksum);
+		float lon = static_cast<float>(mission_item.lon);
+		csum.mission_checksum = crc32part(reinterpret_cast<uint8_t *>(&lon), sizeof(lon), csum.mission_checksum);
+		float alt = mission_item.params[6];
+		csum.mission_checksum = crc32part(reinterpret_cast<uint8_t *>(&alt), sizeof(alt), csum.mission_checksum);
```

Comment:
What about using a `union` where the frame is just `uint8_t[]` instead of all the `reinterpret_cast`ing? I think that would be nicer.

Reply:
You're right - now it looks much nicer!